Add package cache to EmbeddedImplClassLoader

ChrisHegarty opened the pull request:
The initial implementation of the embedded class loader took a brute
force approach to supporting multi-release JARs - iterating over all
possible release versions when searching for classes and resources. This
change improves upon that approach by deriving and caching package and
version specific maps, so class and resource loading can go directly to
the class and resource bytes, respectively, rather than searching.

It's hard to get empirical numbers to quanify just how much this change
improves the performance of classes loaded by this loader, and there is
typically only a couple of hundred classes loaded, but the initial cli
seems observably much quicker, while the server startup has improved
just a bit (at least on my machine).

ChrisHegarty committed:
Add package lookup to Embedded impl classloader

ChrisHegarty committed:
Merge branch 'master' into embed_loader_pkgs

ChrisHegarty committed:
it's privileged dummy!

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

ChrisHegarty committed:
add guava test dep to evil tests - existing issue tickled by this change

ChrisHegarty committed:
do not delegate for packages known to our loader

ChrisHegarty committed:
add small test for CNFE

ChrisHegarty commented:
A very crude measure, printing a timestamp at the beginning of the CLI main and again at the end, then running `elasticsearch --demonize` shows an approximate startup time of 15 seconds before this change and 8.7 seconds with this change, on my Mac M1 (YMMV).

rjernst commented on libs/core/src/main/java/org/elasticsearch/core/internal/provider/EmbeddedImplClassLoader.java:
```diff
@@ -319,14 +428,15 @@ private static Map<JarMeta, CodeSource> getProviderPrefixes(ClassLoader parent,
+                Path p = embeddedJarPath(Set.of(jarPrefix), rootURI)[0];
+                var scan = scanPackages(p);
+                if (p.getFileSystem().provider().getScheme().equals("jar")) {
+                    p.getFileSystem().close();
```
do we need to close this in the case of an exception, ie in tests?

ChrisHegarty replied:
Done.

rjernst commented on qa/evil-tests/build.gradle:
```diff
@@ -17,6 +17,7 @@ apply plugin: 'elasticsearch.standalone-test'
 
 dependencies {
   testImplementation "com.google.jimfs:jimfs:1.2"
+  testImplementation "com.google.guava:guava:${versions.jimfs_guava}"
```
is this needed? I don't see any changes to these tests?

ChrisHegarty replied:
The use of NIO `Path` earlier in some early ES code paths triggers non-default file system providers to be searched, which finds the Jimfs provider, whose static initializer tries to load some Guava classes. I think that this was a pre-existing issue - one cannot load Jimfs without Guava, but was benign if file system providers are not searched.

rjernst reviewed (commented):
Thanks for addressing this perf issue! I have a few thoughts

ChrisHegarty commented on libs/core/src/main/java/org/elasticsearch/core/internal/provider/EmbeddedImplClassLoader.java:
```diff
@@ -305,6 +349,71 @@ static URI rootURI(URL url) {
+                            if (version <= RUNTIME_VERSION_FEATURE) {
+                                String pkg = removeMRJARPrefix(parent).replace(separator, ".");
+                                if (isPackageName(pkg)) {
+                                    pkgVersions.computeIfAbsent(pkg, k -> new ArrayList<>()).add(version);
```
yeah, the versions need to be sorted, which happens a little lower down:

```
 .stream() // sort MR JAR prefixes by release number descending
    .collect(toUnmodifiableMap(Map.Entry::getKey, v -> v.getValue().stream().sorted(Comparator.reverseOrder()).toList()))
        );
```

rjernst replied:
Ah thanks, completely missed that.

ChrisHegarty commented on libs/core/src/main/java/org/elasticsearch/core/internal/provider/EmbeddedImplClassLoader.java:
```diff
@@ -305,6 +349,71 @@ static URI rootURI(URL url) {
+        Map<String, List<Integer>> pkgVersions = new HashMap<>();
+        try {
+            try (var paths = Files.find(dir, Integer.MAX_VALUE, ((path, attrs) -> attrs.isRegularFile()))) {
+                paths.map(path -> dir.relativize(path)).filter(path -> path.getFileName().toString().endsWith(".class")).forEach(path -> {
```
Resources can be versioned too, but I'm not sure that anyone is using that. I deliberately narrowed this change to classes/packages only, and left resources as is (scanning all possible versions), since resources are less of an issue. Also, from a modules perspective, pkgs and directories with non-class files are two different things. I have yet to find a good compromise to allow for both.

ChrisHegarty replied:
Oh, I see - "For non MR packages we can just iterate directories" - Yes, we could.  The set of packages is quite small, and the pkg -> jarMeta is avoids poking about in search. Also, given that we load these classes within modules we can, and should, not delegate for any classes in packages known to this loader.

ChrisHegarty committed:
review comments:
1. assert no prior pkg entries
2. getOrDefault rather than null
3. try try simplify
4. finally close zip file system

ChrisHegarty commented on libs/core/src/main/java/org/elasticsearch/core/internal/provider/EmbeddedImplClassLoader.java:
```diff
@@ -305,6 +349,71 @@ static URI rootURI(URL url) {
+        Set<String> pkgs = new HashSet<>();
+        Map<String, List<Integer>> pkgVersions = new HashMap<>();
+        try {
+            try (var paths = Files.find(dir, Integer.MAX_VALUE, ((path, attrs) -> attrs.isRegularFile()))) {
```
Done.

ChrisHegarty commented on libs/core/src/main/java/org/elasticsearch/core/internal/provider/EmbeddedImplClassLoader.java:
```diff
@@ -125,19 +139,46 @@ public Class<?> findClass(String moduleName, String name) {
+        return AccessController.doPrivileged((PrivilegedAction<Resource>) () -> {
+            List<Integer> releaseVersions = jarMeta.pkgToVersions().get(pkg);
+            if (jarMeta.isMultiRelease() && releaseVersions != null) {
+                for (int releaseVersion : releaseVersions) {
```
Done.

ChrisHegarty commented on libs/core/src/main/java/org/elasticsearch/core/internal/provider/EmbeddedImplClassLoader.java:
```diff
@@ -96,6 +105,11 @@ private EmbeddedImplClassLoader(ClassLoader parent, Map<JarMeta, CodeSource> pre
             .collect(toUnmodifiableMap(k -> k.getKey().prefix(), Map.Entry::getValue));
+        Map<String, JarMeta> map = new HashMap<>();
+        for (var jarMeta : prefixToCodeBase.keySet()) {
+            jarMeta.packages().stream().forEach(pkg -> map.put(pkg, jarMeta));
```
Done.

ChrisHegarty committed:
remove superfluous isMultiRelease check

ChrisHegarty commented on libs/core/src/main/java/org/elasticsearch/core/internal/provider/EmbeddedImplClassLoader.java:
```diff
@@ -125,19 +139,46 @@ public Class<?> findClass(String moduleName, String name) {
+    private Resource privilegedGetClassResourceOrNull(JarMeta jarMeta, String pkg, String filepath) {
+        return AccessController.doPrivileged((PrivilegedAction<Resource>) () -> {
+            List<Integer> releaseVersions = jarMeta.pkgToVersions().get(pkg);
+            if (jarMeta.isMultiRelease() && releaseVersions != null) {
```
Done.

ChrisHegarty committed:
Update resources

ChrisHegarty committed:
Merge branch 'master' into embed_loader_pkgs

rjernst reviewed (approved):
LGTM

ChrisHegarty committed:
make variable local

ChrisHegarty commented on libs/core/src/main/java/org/elasticsearch/core/internal/provider/EmbeddedImplClassLoader.java:
```diff
@@ -128,50 +143,94 @@ public Class<?> findClass(String moduleName, String name) {
+        if (jarMeta != null) {
+            url = findResourceInLoaderPkgOrNull(jarMeta, pkg, name, parent::getResource);
+        } else {
+            // not in a package defined to this loader, slow search
```
> I guess it is for things like META-INF or any resources at the root?
Correct.

> Are those even "versionable" with MRJARs? 
Those are not, but non-package name directories are versionable, e.g. `/a-b/resource.txt`. The test updates asserts this, and also compares the embedded impl loader behaviour to that of URLClassLoader.

ChrisHegarty committed:
Merge branch 'main' into embed_loader_pkgs

ChrisHegarty commented on libs/core/src/main/java/org/elasticsearch/core/internal/provider/EmbeddedImplClassLoader.java:
```diff
@@ -128,50 +143,94 @@ public Class<?> findClass(String moduleName, String name) {
-    <T> T findResourceForPrefixOrNull(String name, JarMeta jarMeta, Function<String, T> finder) {
-        T resource;
+    URL findResourceForPrefixOrNull(String name, JarMeta jarMeta) {
+        URL url;
```
Done.
